src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -245,85 +245,88 @@ void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp,
   sample.set_timestamp_ns(event_timestamp);
 
   Callstack* callstack = sample.mutable_callstack();
+  callstack->set_type(Callstack::kComplete);
+
+  bool has_unwinding_error = false;
 
   // Callstacks with only two frames (the first is in the kernel, the second is the sampled address)
   // are unwinding errors.
   // Note that this doesn't exclude samples inside the main function of any thread as the main
   // function is never the outermost frame. For example, for the main thread the outermost function
   // is _start, followed by __libc_start_main. For other threads, the outermost function is clone.
-  if (event_data.GetCallchainSize() == 2) {
+  if (!has_unwinding_error && event_data.GetCallchainSize() == 2) {
     if (unwind_error_counter_ != nullptr) {
       ++(*unwind_error_counter_);
     }
     callstack->set_type(Callstack::kFramePointerUnwindingError);
-    callstack->add_pcs(event_data.GetCallchain()[1]);
-    listener_->OnCallstackSample(std::move(sample));
-    return;
+    has_unwinding_error = true;
   }
 
   uint64_t top_ip = event_data.GetCallchain()[1];
 
   // Some samples can actually fall inside u(ret)probes code. Set their type accordingly, as we
   // don't want to show the unnamed uprobes module in the samples.
   unwindstack::MapInfo* top_ip_map_info = current_maps_->Find(top_ip);
-  if (top_ip_map_info != nullptr && top_ip_map_info->name() == "[uprobes]") {
+  if (!has_unwinding_error && top_ip_map_info != nullptr &&
+      top_ip_map_info->name() == "[uprobes]") {
     if (samples_in_uretprobes_counter_ != nullptr) {
       ++(*samples_in_uretprobes_counter_);
     }
     callstack->set_type(Callstack::kInUprobes);
-    callstack->add_pcs(top_ip);
-    listener_->OnCallstackSample(std::move(sample));
-    return;
+    has_unwinding_error = true;
```

Comment:
I don't know if we really want to show all frames in all cases here. Some of these are very different. My understanding is that showing all frames could provide value in the case of failed leaf function patching and in the case of general frame pointer unwinding error (e.g., this can happen after multiple correct frames). But I'm not sure it does in other cases like this one (uprobes) and the user space instrumentation ones. Also we would still show no extra frame in the case of DWARF errors and so it would be consistent to only pick a few cases we think are interesting.

Reply:
As discussed. I actually forgot the Dwarf case. I think we always want to expose any information we have. It will not confuse the user, as the information is marked as "unwinding error".